cli/ops/worker_host.rs

```diff
@@ -37,6 +44,190 @@ pub fn init(i: &mut Isolate, s: &ThreadSafeState) {
+      ErrorKind::JSError => {
+        let error = error.downcast::<JSError>().unwrap();
+        let exception: V8Exception = error.into();
+        json!({"error": {
+          "message": exception.message,
+          "fileName": exception.script_resource_name,
+          "lineNumber": exception.line_number,
+          "columnNumber": exception.start_column,
+        }})
+      }
+      _ => json!({"error": {
+        "message": error.to_string(),
+      }}),
+    }
+  } else {
+    json!({"ok": true})
+  }
+}
+
+fn create_web_worker(
+  name: String,
+  global_state: GlobalState,
+  permissions: Arc<Mutex<DenoPermissions>>,
+  specifier: ModuleSpecifier,
+) -> Result<WebWorker, ErrBox> {
+  let state = ThreadSafeState::new_for_worker(
+    global_state,
+    Some(permissions),
+    specifier,
+  )?;
+
+  let mut worker =
+    WebWorker::new(name.to_string(), startup_data::deno_isolate_init(), state);
+
+  // 2. Bootstrap runtime
+  let script = format!("bootstrapWorkerRuntime(\"{}\")", name);
+  worker.execute(&script)?;
+
+  Ok(worker)
+}
+
+// TODO(bartlomieju): check if order of actions is aligned to Worker spec
+fn run_worker_thread(
+  name: String,
+  global_state: GlobalState,
+  permissions: Arc<Mutex<DenoPermissions>>,
+  specifier: ModuleSpecifier,
+  has_source_code: bool,
+  source_code: String,
+) -> Result<WorkerChannelsExternal, ErrBox> {
+  let (handle_sender, handle_receiver) =
+    std::sync::mpsc::sync_channel::<Result<WorkerChannelsExternal, ErrBox>>(1);
+
+  // TODO(bartlomieju): should we store JoinHandle as well?
+  std::thread::spawn(move || {
+    // Any error inside this block is terminal:
+    // - JS worker is useless - meaning it throws an exception and can't do anything else,
+    //  all action done upon it should be noops
+    // - newly spawned thread exits
+    let result =
+      create_web_worker(name, global_state, permissions, specifier.clone());
+
+    if let Err(err) = result {
+      handle_sender.send(Err(err)).unwrap();
+      return;
+    }
+
+    let mut worker = result.unwrap();
+    // Send thread safe handle to newly created worker to host thread
+    handle_sender.send(Ok(worker.thread_safe_handle())).unwrap();
+    drop(handle_sender);
+
+    // At this point the only method of communication with host
+    // is using `state.internal_worker_channels`.
+    //
+    // Host can already push messages and interact with worker.
+    //
+    // Next steps:
+    // - create tokio runtime
+    // - load provided module or code
+    // - start driving worker's event loop
+
+    let mut rt = create_basic_runtime();
+    // Execute provided source code immediately
+    let result = if has_source_code {
+      worker.execute(&source_code)
+    } else {
+      // TODO(bartlomieju): add "type": "classic", ie. ability to load
+      // script instead of module
+      let load_future = worker
+        .execute_mod_async(&specifier, None, false)
+        .boxed_local();
+
+      rt.block_on(load_future)
```

Comment:
This should be `select!()` between module execution and termination - https://html.spec.whatwg.org/multipage/workers.html#worker-processing-model - point 24